tensorflow/compiler/tf2tensorrt/convert/ops/layer_utils.h

```diff
@@ -0,0 +1,354 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+#ifndef TENSORFLOW_COMPILER_TF2TENSORRT_CONVERT_OPS_LAYER_UTILS_H_
+#define TENSORFLOW_COMPILER_TF2TENSORRT_CONVERT_OPS_LAYER_UTILS_H_
+#if GOOGLE_CUDA && GOOGLE_TENSORRT
+
+#include <type_traits>
+
+#include "tensorflow/compiler/tf2tensorrt/convert/convert_nodes.h"
+#include "tensorflow/compiler/tf2tensorrt/convert/utils.h"
+#include "tensorflow/core/framework/types.h"
+#include "tensorflow/core/platform/statusor.h"
+#include "third_party/tensorrt/NvInfer.h"
+#include "third_party/tensorrt/NvInferRuntimeCommon.h"
+
+namespace tensorflow {
+namespace tensorrt {
+
+namespace convert {
+
+// Simplifies the creation of a TensorRT network. The user provides a
+// INetworkDefinition pointer during construction. They can then add operations
+// to the network through the provided functions. Each function returns a struct
+// which contains the symbolic result of the operation (ITensor pointer) as well
+// as a pointer to the last TensorRT ILayer created. Some operations may create
+// multiple layers in order to accomplish the desired result (e.g. Sign).
+class TRTNetworkBuilder {
+ public:
+  TRTNetworkBuilder(nvinfer1::INetworkDefinition* network,
+                    TrtWeightStore* weight_store)
+      : network_(network), weight_store_(weight_store) {}
+
+  template <typename T, typename std::enable_if<std::is_base_of<
+                            nvinfer1::ILayer, T>::value>::type* = nullptr>
+  struct CreationResult {
+    T* layer;
+    nvinfer1::ITensor* output;
+  };
+
+  // Adds an Add operation to the network.
+  StatusOr<nvinfer1::IElementWiseLayer*> Add(nvinfer1::ITensor* lhs,
+                                             nvinfer1::ITensor* rhs) noexcept {
+    TRT_EXPECT(network_);
+    TRT_EXPECT(lhs);
+    TRT_EXPECT(rhs);
+    auto layer = network_->addElementWise(*lhs, *rhs,
+                                          nvinfer1::ElementWiseOperation::kSUM);
+    return MakeResult(layer);
+  };
+
+  // Adds an Add operation to the network. Note that this unary operations will
+  // do an implict float conversion.
+  StatusOr<nvinfer1::IUnaryLayer*> Abs(nvinfer1::ITensor* input) noexcept {
+    TRT_EXPECT(network_);
+    TRT_EXPECT(input);
+    auto layer = network_->addUnary(*input, nvinfer1::UnaryOperation::kABS);
+    return MakeResult(layer);
+  }
+
+  // Performs Abs without implict float conversion.
+  StatusOr<nvinfer1::IElementWiseLayer*> AbsInt(
+      nvinfer1::ITensor* input) noexcept {
+    TRT_EXPECT(network_);
+    TRT_EXPECT(input);
+    auto sign = this->Sign(input);
+    TRT_ENSURE_OK(sign);
+    return Mul(input, (*sign)->getOutput(0));
```

Comment:
To follow up with my comment about TRT_ENSURE_OK and TRT_ENSURE_PTR_OR, here why don't we use TRT_ENSURE_PTR_OK? Do we assume if sign has a pointer then the pointer is not null?

Reply:
Internally it's already checked by the routine that produced "sign"

Reply:
I changed to TRT_ENSURE_PTR_OK for consistency